testtime/testtime.go

```diff
@@ -0,0 +1,213 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package testtime provides a mocked version of time.Timer for use in tests.
+package testtime
+
+import (
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/timeutil"
+)
+
+// TestTimer is a mocked version of time.Timer for which the passage of time or
+// the direct expiration of the timer is controlled manually.
+//
+// TestTimer implements timeutil.Timer.
+//
+// TestTimer also provides methods to introspect whether the timer is active or
+// how many times it has fired.
+type TestTimer struct {
+	lock       sync.Mutex
+	currTime   time.Time
+	expiration time.Time
+	active     bool
+	fireCount  int
+	callback   func()
+	c          chan time.Time
+}
+
+var _ timeutil.Timer = (*TestTimer)(nil)
+
+// AfterFunc waits for the timer to fire and then calls f in its own goroutine.
+// It returns a Timer that can be used to cancel the call using its Stop method.
+// The returned Timer's C field is not used and will be nil.
+//
+// AfterFunc returns a TestTimer which simulates the behavior of a timer which
+// was created via time.AfterFunc.
+//
+// See here for more details: https://pkg.go.dev/time#AfterFunc
+func AfterFunc(d time.Duration, f func()) *TestTimer {
+	osutil.MustBeTestBinary("testtime timers cannot be used outside of tests")
+	currTime := time.Now()
+	timer := &TestTimer{
+		currTime:   currTime,
+		expiration: currTime.Add(d),
+		active:     true,
+		callback:   f,
+	}
+	if d <= 0 {
+		defer timer.doFire(timer.currTime)
+	}
+	return timer
+}
+
+// NewTimer creates a new Timer that will send the current time on its channel
+// after the timer fires.
+//
+// NewTimer returns a TestTimer which simulates the behavior of a timer which
+// was created via time.NewTimer.
+//
+// See here for more details: https://pkg.go.dev/time#NewTimer
+func NewTimer(d time.Duration) *TestTimer {
+	osutil.MustBeTestBinary("testtime timers cannot be used outside of tests")
+	currTime := time.Now()
```

Comment:
same remark, in other places where we mock time, we usually do so by mocking exactly timeNow